File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stack;
        for (char c: s) {
            if (c == '(' or c == '{' or c == '[') stack.push(c);
            else {
                if (stack.empty() 
                    or (stack.top() == '{' and c != '}') 
                    or (stack.top() == '(' and c != ')')
                    or (stack.top() == '[' and c != ']')) return false;
                stack.pop();
            }
        }
        return stack.empty();
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.kt
class Solution {
    fun isValid(s: String): Boolean {
        var stack = Stack<Char>()
        for (c: Char in s) {
            when {
                c == '(' -> stack.push(')')
                c == '{' -> stack.push('}')
                c == '[' -> stack.push(']')
                (stack.isEmpty() || stack.pop() != c) -> return false
            }
        }
        return stack.isEmpty()
    }
}